### ISSUE TITLE
\Roave\BetterReflection\Reflection\ReflectionClass "Carbon\Traits\Localization" could not be found in the located source

### ISSUE THREAD

**Issue by Aerendir** [Roave/BetterReflection#689]
As per object, the error returned is this:

> [ERROR] Could not process "src/Strategy/TimeFixedStrategy.php" file, due to:                                           
         "_HumbugBox10beb0db089e\Roave\BetterReflection\Reflection\ReflectionClass "Carbon\Traits\Localization" could   
         not be found in the located source".

The trait in question [is this](https://github.com/briannesbitt/Carbon/blob/master/src/Carbon/Traits/Localization.php).

**Comment by ondrejmirtes** [comment]
What project are you using? Seems like the BetterReflection is packed in a PHAR, just interested when you actually see the error.

Anyway, this problem can be due to misconfigured source locators that actually don't see the trait, so the error might be expected.

**Comment by Ocramius** [comment]
Can this be reproduced in isolation with some steps?

**Comment by Aerendir** [comment]
Excuse me for the late response: I missed it...

Anyway, the error happens on this project: https://github.com/Aerendir/component-then-when

# Steps to reproduce

1. Obviously, clone the repo locally
2. In file `rector.php`, comment the line that excludes the file `__DIR__ . '/src/Strategy/TimeFixedStrategy.php'`
3. Run `bin/validate`

```diff
// rector.php, line 30

...
    $parameters->set(Option::EXCLUDE_PATHS, [
        __DIR__ . '/tests/bootstrap.php',
-        __DIR__ . '/src/Strategy/TimeFixedStrategy.php',
    ]);
...

```

**Comment by AntiPaste** [comment]
I'm also encountering this issue. I'm not sure if `BetterReflection` is the problem or if this should be reported to https://github.com/rector/rector or perhaps something in-between like https://github.com/phpstan/phpstan which is actually the library using `BetterReflection`.

Here is my minimal reproduction with just `rector/rector` and `briannesbitt/carbon` installed: https://github.com/AntiPaste/better-reflection-test

Full stack trace of the error:
```plain
In IdentifierNotFound.php line 24:
                                                                                                            
  [_HumbugBoxf43f7c5c5350\Roave\BetterReflection\Reflector\Exception\IdentifierNotFound]                    
  _HumbugBoxf43f7c5c5350\Roave\BetterReflection\Reflection\ReflectionClass "Carbon\Traits\Localization" co  
  uld not be found in the located source                                                                    
                                                                                                            

Exception trace:
  at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/vendor/ondrejmirtes/better-reflection/src/Reflector/Exception/IdentifierNotFound.php:24
 _HumbugBoxf43f7c5c5350\Roave\BetterReflection\Reflector\Exception\IdentifierNotFound::fromIdentifier() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/vendor/ondrejmirtes/better-reflection/src/Reflector/ClassReflector.php:47
 _HumbugBoxf43f7c5c5350\Roave\BetterReflection\Reflector\ClassReflector->reflect() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/BetterReflection/Reflector/MemoizingClassReflector.php:29
 PHPStan\Reflection\BetterReflection\Reflector\MemoizingClassReflector->reflect() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/vendor/ondrejmirtes/better-reflection/src/Reflection/ReflectionClass.php:731
 _HumbugBoxf43f7c5c5350\Roave\BetterReflection\Reflection\ReflectionClass->reflectClassForNamedNode() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/vendor/ondrejmirtes/better-reflection/src/Reflection/ReflectionClass.php:715
 _HumbugBoxf43f7c5c5350\Roave\BetterReflection\Reflection\ReflectionClass->_HumbugBoxf43f7c5c5350\Roave\BetterReflection\Reflection\{closure}() at n/a:n/a
 array_map() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/vendor/ondrejmirtes/better-reflection/src/Reflection/ReflectionClass.php:720
 _HumbugBoxf43f7c5c5350\Roave\BetterReflection\Reflection\ReflectionClass->getTraits() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/vendor/ondrejmirtes/better-reflection/src/Reflection/ReflectionClass.php:232
 _HumbugBoxf43f7c5c5350\Roave\BetterReflection\Reflection\ReflectionClass->getMethodsFromTraits() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/vendor/ondrejmirtes/better-reflection/src/Reflection/ReflectionClass.php:264
 _HumbugBoxf43f7c5c5350\Roave\BetterReflection\Reflection\ReflectionClass->getMethodsIndexedByName() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/vendor/ondrejmirtes/better-reflection/src/Reflection/ReflectionClass.php:309
 _HumbugBoxf43f7c5c5350\Roave\BetterReflection\Reflection\ReflectionClass->getMethods() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/vendor/ondrejmirtes/better-reflection/src/Reflection/ReflectionClass.php:231
 _HumbugBoxf43f7c5c5350\Roave\BetterReflection\Reflection\ReflectionClass->_HumbugBoxf43f7c5c5350\Roave\BetterReflection\Reflection\{closure}() at n/a:n/a
 array_map() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/vendor/ondrejmirtes/better-reflection/src/Reflection/ReflectionClass.php:232
 _HumbugBoxf43f7c5c5350\Roave\BetterReflection\Reflection\ReflectionClass->getMethodsFromTraits() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/vendor/ondrejmirtes/better-reflection/src/Reflection/ReflectionClass.php:264
 _HumbugBoxf43f7c5c5350\Roave\BetterReflection\Reflection\ReflectionClass->getMethodsIndexedByName() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/vendor/ondrejmirtes/better-reflection/src/Reflection/ReflectionClass.php:309
 _HumbugBoxf43f7c5c5350\Roave\BetterReflection\Reflection\ReflectionClass->getMethods() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/vendor/ondrejmirtes/better-reflection/src/Reflection/ReflectionClass.php:231
 _HumbugBoxf43f7c5c5350\Roave\BetterReflection\Reflection\ReflectionClass->_HumbugBoxf43f7c5c5350\Roave\BetterReflection\Reflection\{closure}() at n/a:n/a
 array_map() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/vendor/ondrejmirtes/better-reflection/src/Reflection/ReflectionClass.php:232
 _HumbugBoxf43f7c5c5350\Roave\BetterReflection\Reflection\ReflectionClass->getMethodsFromTraits() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/vendor/ondrejmirtes/better-reflection/src/Reflection/ReflectionClass.php:264
 _HumbugBoxf43f7c5c5350\Roave\BetterReflection\Reflection\ReflectionClass->getMethodsIndexedByName() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/vendor/ondrejmirtes/better-reflection/src/Reflection/ReflectionClass.php:309
 _HumbugBoxf43f7c5c5350\Roave\BetterReflection\Reflection\ReflectionClass->getMethods() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/vendor/ondrejmirtes/better-reflection/src/Reflection/ReflectionClass.php:220
 _HumbugBoxf43f7c5c5350\Roave\BetterReflection\Reflection\ReflectionClass::_HumbugBoxf43f7c5c5350\Roave\BetterReflection\Reflection\{closure}() at n/a:n/a
 array_map() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/vendor/ondrejmirtes/better-reflection/src/Reflection/ReflectionClass.php:221
 _HumbugBoxf43f7c5c5350\Roave\BetterReflection\Reflection\ReflectionClass->getParentMethods() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/vendor/ondrejmirtes/better-reflection/src/Reflection/ReflectionClass.php:263
 _HumbugBoxf43f7c5c5350\Roave\BetterReflection\Reflection\ReflectionClass->getMethodsIndexedByName() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/vendor/ondrejmirtes/better-reflection/src/Reflection/ReflectionClass.php:346
 _HumbugBoxf43f7c5c5350\Roave\BetterReflection\Reflection\ReflectionClass->getMethod() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/vendor/ondrejmirtes/better-reflection/src/Reflection/ReflectionClass.php:358
 _HumbugBoxf43f7c5c5350\Roave\BetterReflection\Reflection\ReflectionClass->hasMethod() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/vendor/ondrejmirtes/better-reflection/src/Reflection/Adapter/ReflectionClass.php:161
 _HumbugBoxf43f7c5c5350\Roave\BetterReflection\Reflection\Adapter\ReflectionClass->hasMethod() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/Php/PhpClassReflectionExtension.php:208
 PHPStan\Reflection\Php\PhpClassReflectionExtension->hasMethod() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/ClassReflection.php:266
 PHPStan\Reflection\ClassReflection->hasMethod() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/src/Type/ObjectType.php:303
 PHPStan\Type\ObjectType->hasMethod() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php:2547
 PHPStan\Analyser\MutatingScope->methodCallReturnType() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php:1218
 PHPStan\Analyser\MutatingScope->PHPStan\Analyser\{closure}() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/src/Type/TypeTraverser.php:47
 PHPStan\Type\TypeTraverser->mapInternal() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/src/Type/TypeTraverser.php:37
 PHPStan\Type\TypeTraverser::map() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php:1220
 PHPStan\Analyser\MutatingScope->resolveType() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php:372
 PHPStan\Analyser\MutatingScope->getType() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php:937
 PHPStan\Analyser\MutatingScope->resolveType() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php:372
 PHPStan\Analyser\MutatingScope->getType() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php:683
 PHPStan\Analyser\MutatingScope->resolveType() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php:372
 PHPStan\Analyser\MutatingScope->getType() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/ParametersAcceptorSelector.php:37
 PHPStan\Reflection\ParametersAcceptorSelector::selectFromArgs() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php:2572
 PHPStan\Analyser\MutatingScope->methodCallReturnType() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php:1218
 PHPStan\Analyser\MutatingScope->PHPStan\Analyser\{closure}() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/src/Type/TypeTraverser.php:47
 PHPStan\Type\TypeTraverser->mapInternal() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/src/Type/TypeTraverser.php:37
 PHPStan\Type\TypeTraverser::map() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php:1220
 PHPStan\Analyser\MutatingScope->resolveType() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php:372
 PHPStan\Analyser\MutatingScope->getType() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php:1058
 PHPStan\Analyser\NodeScopeResolver->PHPStan\Analyser\{closure}() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php:1784
 PHPStan\Analyser\NodeScopeResolver->processAssignVar() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php:1067
 PHPStan\Analyser\NodeScopeResolver->processExprNode() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php:372
 PHPStan\Analyser\NodeScopeResolver->processStmtNode() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php:224
 PHPStan\Analyser\NodeScopeResolver->processStmtNodes() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php:342
 PHPStan\Analyser\NodeScopeResolver->processStmtNode() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php:224
 PHPStan\Analyser\NodeScopeResolver->processStmtNodes() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php:414
 PHPStan\Analyser\NodeScopeResolver->processStmtNode() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php:224
 PHPStan\Analyser\NodeScopeResolver->processStmtNodes() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php:383
 PHPStan\Analyser\NodeScopeResolver->processStmtNode() at phar:///truncated/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php:193
 PHPStan\Analyser\NodeScopeResolver->processNodes() at /truncated/vendor/rector/rector/packages/node-type-resolver/src/PHPStan/Scope/PHPStanNodeScopeResolver.php:158
 Rector\NodeTypeResolver\PHPStan\Scope\PHPStanNodeScopeResolver->processNodes() at /truncated/vendor/rector/rector/packages/node-type-resolver/src/NodeScopeAndMetadataDecorator.php:123
 Rector\NodeTypeResolver\NodeScopeAndMetadataDecorator->decorateNodesFromFile() at /truncated/vendor/rector/rector/src/Application/FileProcessor.php:172
 Rector\Core\Application\FileProcessor->parseAndTraverseFileInfoToNodes() at /truncated/vendor/rector/rector/src/Application/FileProcessor.php:105
 Rector\Core\Application\FileProcessor->parseFileInfoToLocalCache() at /truncated/vendor/rector/rector/src/Application/RectorApplication.php:215
 Rector\Core\Application\RectorApplication->Rector\Core\Application\{closure}() at /truncated/vendor/rector/rector/src/Application/RectorApplication.php:242
 Rector\Core\Application\RectorApplication->tryCatchWrapper() at /truncated/vendor/rector/rector/src/Application/RectorApplication.php:216
 Rector\Core\Application\RectorApplication->parseFileInfosToNodes() at /truncated/vendor/rector/rector/src/Application/RectorApplication.php:135
 Rector\Core\Application\RectorApplication->runOnFileInfos() at /truncated/vendor/rector/rector/src/Console/Command/ProcessCommand.php:235
 Rector\Core\Console\Command\ProcessCommand->execute() at /truncated/vendor/symfony/console/Command/Command.php:258
 Symfony\Component\Console\Command\Command->run() at /truncated/vendor/rector/rector/src/Console/Command/AbstractCommand.php:35
 Rector\Core\Console\Command\AbstractCommand->run() at /truncated/vendor/symfony/console/Application.php:920
 Symfony\Component\Console\Application->doRunCommand() at /truncated/vendor/symfony/console/Application.php:266
 Symfony\Component\Console\Application->doRun() at /truncated/vendor/rector/rector/src/Console/Application.php:100
 Rector\Core\Console\Application->doRun() at /truncated/vendor/symfony/console/Application.php:142
 Symfony\Component\Console\Application->run() at /truncated/vendor/rector/rector/bin/rector:82
```

**Comment by Ocramius** [comment]
Probably interesting to look at how `rector` or `phpstan` are initializing the source locators.

Once we have that, we can reproduce it in isolation with just this library, which would expose the actual problem.

**Comment by AntiPaste** [comment]
Quick update: downgrading `phpstan/phpstan` to 0.12.42 and `nikic/php-parser` to v4.9.0 solves the problem. `phpstan/phpstan` 0.12.43 requires `nikic/php-parser` >= 4.10.0 so it's difficult to say which one is actually causing the issue, but it seems like it's not this project.

**Comment by Ocramius** [comment]
@AntiPaste did the `roave/better-reflection` version change at all?

**Comment by Aerendir** [comment]
@TomasVotruba , please, can you have a look here?

**Comment by Ocramius** [comment]
Eh, folks, I know @TomasVotruba is a helpful person, but this really needs your input/debugging first :P 

Specifically, in the phar you are using, look for where better-reflection is being instantiated.

**Comment by Aerendir** [comment]
@Ocramius , the fact is that I don't use any Phar 😅 Fr this reason I don't know from where the error comes.

I'm using Bamarni to manage dev tools, but at this point I don't think it is the root cause: @AntiPaste, are you using `bamarni/composer-bin-plugin`?

For me debuggin this issue is a really complex task: here we have four packages involved: Rector, PHPStan, BetterReflection and, maybe, ComposerBinPlugin.

And I don't know those packages enough to debug them without spending days only to understand their mechanics...

@AntiPaste , I've tried your test package (https://github.com/AntiPaste/better-reflection-test):

```console
git clone git@github.com:AntiPaste/better-reflection-test.git
composer install
vendor/bin/rector process
```

No errors: how do you use it?

**Comment by paali** [comment]
Just a quick update (maybe not yet helpful). It seems that the problem is that `\PHPStan\Reflection\BetterReflection\SourceLocator\AutoloadSourceLocator::locateClassByName` resolves the line number for the Localization trait as 36. But `\PhpParser\NodeAbstract` resolves the line number for the Trait as 24.

```
<?php

/**
 * This file is part of the Carbon package.
 *
 * (c) Brian Nesbitt <brian@nesbot.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
namespace Carbon\Traits;

use Carbon\CarbonInterface;
use Carbon\Exceptions\InvalidTypeException;
use Carbon\Exceptions\NotLocaleAwareException;
use Carbon\Language;
use Carbon\Translator;
use Closure;
use Symfony\Component\Translation\TranslatorBagInterface;
use Symfony\Component\Translation\TranslatorInterface;
use Symfony\Contracts\Translation\LocaleAwareInterface;
use Symfony\Contracts\Translation\TranslatorInterface as ContractsTranslatorInterface;

if (!interface_exists('Symfony\\Component\\Translation\\TranslatorInterface')) {  // <-- This here is 24
    class_alias(
        'Symfony\\Contracts\\Translation\\TranslatorInterface',
        'Symfony\\Component\\Translation\\TranslatorInterface'
    );
}

/**
 * Trait Localization.
 *
 * Embed default and locale translators and translation base methods.
 */
trait Localization // <-- This here is 36
```
I haven't find out yet who sets up the NodeAbstract linenumber. I'll try to continue when I find some more time.

Any ideas where these findings belong to? nikic/php-parse seems to be the one that comes into wrong conclusion about the line number.

**Comment by paali** [comment]
Removing the class aliasing also seems to make the problem disappear.

**Comment by king40or1** [comment]
How can I update

**Comment by Ocramius** [comment]
This should be solved by #802 - needs checking, but beware that it will be a major version upgrade.

**Comment by Ocramius** [comment]
Closing: marking this as solved, even if I don't have the time to verify it myself.

Please consider opening a new issue after upgrading to `5.0.0`, which I will release in the next few days.
